### PR TITLE
feat(context): add context support

### PR DESCRIPTION
**Pull request by yeqown**

### CODE REVIEW

**Comment by yeqown** [comment]
Resolve https://github.com/nats-io/nats-top/issues/97

**Comment by yeqown** [comment]
@wallyqs @dsidirop

**Comment by jnmoyne** [comment]
The only issue I see here is that the context doesn't store which http/https monitoring port should be used. 
Wouldn't it be better to create a PR on jsm.go first that adds HTTP/S monitoring ports to the context, and then this PR which would be amended to get the monitoring port number from the context.

**Comment by wallyqs** [comment]
Thanks for the contribution, this would be convenient when in k8s scenarios inside of nats-box where it is assumed that the port is 8222 and there is a preconfigured context.  One thing to check too is whether the binary size increases too much by adding the dependency, maybe could borrow too a couple of functions from the natscontext library just to load the current active context hostname since will not be using creds or the nats client for the 8222 port.

**Comment by yeqown** [comment]
> Thanks for the contribution, this would be convenient when in k8s scenarios inside of nats-box where it is assumed that the port is 8222 and there is a preconfigured context. One thing to check too is whether the binary size increases too much by adding the dependency, maybe could borrow too a couple of functions from the natscontext library just to load the current active context hostname since will not be using creds or the nats client for the 8222 port.

Does the binary size matter? As far as I can see, keeping the consistency of the nats-context API in all nat-xxx CLI tools maybe more important. Then I tested it, the result is here:

```sh
➜  bin ls -lh | grep nats-top (new)
-rwxr-xr-x  1 apple  staff   8.4M 12 27 16:07 nats-top
➜  bin ls -lh | grep nats-top (old)
-rwxr-xr-x@ 1 apple  staff   8.3M 12 28 09:34 nats-top
➜  bin
```

The size increment of the import package much cheap than the runtime itself

**Comment by yeqown** [comment]
> The only issue I see here is that the context doesn't store which http/https monitoring port should be used. Wouldn't it be better to create a PR on jsm.go first that adds HTTP/S monitoring ports to the context, and then this PR which would be amended to get the monitoring port number from the context.

I'll try it, but I'm not so sure would it be accepted, but just try it

File: nats-top.go [hunk 1]
```diff
@@ -36,12 +36,15 @@ var (
 	caCertOpt     = flag.String("cacert", "", "Root CA cert")
 	skipVerifyOpt = flag.Bool("k", false, "Skip verifying server certificate")
 
+	// Nats context options
+	contextName = flag.String("c", "", "The name of the context to use. If not specified, the currently selected context is used.")
```
Comment (wallyqs) [inline nats-top.go:40]: would prefer here `--context` like the nats cli too
```suggestion
	contextName = flag.String("context", "", "The name of the context to use. If not specified, the currently selected context is used.")
```

File: nats-top.go [hunk 2]
```diff
@@ -36,12 +36,15 @@ var (
...
+
 	version = "0.0.0"
 )
 
 const usageHelp = `
 usage: nats-top [-s server] [-m http_port] [-ms https_port] [-n num_connections] [-d delay_secs] [-r max] [-o FILE] [-l DELIMITER] [-sort by]
-                [-cert FILE] [-key FILE] [-cacert FILE] [-k] [-b] [-v|--version] [-u|--display-subscriptions-column]
+                [-cert FILE] [-key FILE] [-cacert FILE] [-k] [-b] [-v|--version] [-u|--display-subscriptions-column] [-c context_name]
```
Comment (wallyqs) [inline nats-top.go:47]: same here, let's just use longer option for now
```suggestion
                [-cert FILE] [-key FILE] [-cacert FILE] [-k] [-b] [-v|--version] [-u|--display-subscriptions-column] [--context context_name]
```

File: readme.md [hunk 3]
```diff
@@ -93,6 +93,10 @@ usage: nats-top [-s server] [-m http_port] [-ms https_port] [-n num_connections]
 
   Makes the subscriptions-column immediately visible upon launching nats-top.
 
+- `-c context_name`
```
Comment (wallyqs) [inline readme.md:96]:
```suggestion
- `--context context_name`
```

File: context.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+		name = natscontext.SelectedContext()
+	}
+	if name == "" {
+		fmt.Printf("nats-top: context name required\n")
+		return
+	}
+
+	natsCtx, err := natscontext.New(name, true)
+	if err != nil {
+		fmt.Printf("nats-top: error load context: %s\n", err)
```
Comment (wallyqs) [inline context.go:25]:
```suggestion
		fmt.Fprintf(os.Stderr, "nats-top: error loading context: %s\n", err)
```

File: context.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+)
+
+// withContext will load a context and set it's setting to the global
+// flag values: host, certOpt, keyOpt, caCertOpt
+func withContext(name string) {
+	if name == "" {
+		name = natscontext.SelectedContext()
+	}
+	if name == "" {
+		fmt.Printf("nats-top: context name required\n")
```
Comment (wallyqs) [inline context.go:19]:
```suggestion
		fmt.Fprintf(os.Stderr, "nats-top: context name required\n")
```

**Review by wallyqs — APPROVED** [review]
LGTM, left a few comments but I can apply the changes after merging too. Thanks for the contribution!

**Comment by ripienaar** [comment]
I wouldn’t add the monitor port to the context sorry.  Everything in natscontext id about cluster access. Nats too should become network aware. 

I don’t see setting monitor port as a flag here as a big problem

**Comment by ripienaar** [comment]
> Thanks for the contribution, this would be convenient when in k8s scenarios inside of nats-box where it is assumed that the port is 8222 and there is a preconfigured context. One thing to check too is whether the binary size increases too much by adding the dependency, maybe could borrow too a couple of functions from the natscontext library just to load the current active context hostname since will not be using creds or the nats client for the 8222 port.

Would rather move nats intent out to its own package to trim dependencies. 

But even better is that nats.go implement the context adr

**Comment by ripienaar** [comment]
Rather than (ab)using context for a thing its not needed I suggest we move nats-top to nats instead and deprecate http access to monitoring data.

I have some more testing to do but a import of this code into cli is here and it works

https://github.com/nats-io/natscli/pull/942

**Comment by jnmoyne** [comment]
+100% on nats.go should be implementing the context ADR (so it could move there from jsm.go)

**Comment by wallyqs** [comment]
@yeqown thanks again for the contribution, looks like context usage would be a better fit following this approach which you can try: https://github.com/nats-io/natscli/pull/942 I will close this one for now.

**Comment by yeqown** [comment]
Anyway, there is a good result for users 😀

**Pull request closed without merging**